velox/exec/fuzzer/WindowFuzzer.cpp

```diff
@@ -60,20 +61,148 @@ void WindowFuzzer::addWindowFunctionSignatures(
+      0, startBoundOptions.size() - 1)(rng_);
+  auto endBoundMinIdx = std::max(0, static_cast<int>(startBoundIndex) - 1);
+  auto endBoundIndex = boost::random::uniform_int_distribution<uint32_t>(
+      endBoundMinIdx, endBoundOptions.size() - 1)(rng_);
+  auto frameStartBoundType = startBoundOptions[startBoundIndex];
+  auto frameEndBoundType = endBoundOptions[endBoundIndex];
+
+  return std::make_tuple(windowType, frameStartBoundType, frameEndBoundType);
+}
+
+// For frames with k RANGE PRECEDING/FOLLOWING, Velox requires the application
+// to add columns with the range frame boundary value computed according to the
+// frame type.
+// If the frame is k PRECEDING :
+// frame_boundary_value = current_order_by - k (for ascending ORDER BY)
+// frame_boundary_value = current_order_by + k (for descending ORDER BY)
+// If the frame is k FOLLOWING :
+// frame_boundary_value = current_order_by + k (for ascending ORDER BY)
+// frame_boundary_value = current_order_by - k (for descending ORDER BY)
+template <typename T>
+T WindowFuzzer::genOffsetAtIdx(
+    const T& offsetColumnVal,
+    T offsetValue,
+    core::WindowNode::BoundType frameBoundType,
+    core::SortOrder sortOrder) {
+  if ((frameBoundType == core::WindowNode::BoundType::kPreceding &&
+       sortOrder.isAscending()) ||
+      (frameBoundType == core::WindowNode::BoundType::kFollowing &&
+       !sortOrder.isAscending())) {
+    if constexpr (std::is_same_v<T, double> || std::is_same_v<T, float>) {
+      return offsetColumnVal - offsetValue;
+    } else {
+      return checkedMinus<T>(offsetColumnVal, offsetValue);
     }
-  };
-  auto isRowsFrame =
-      boost::random::uniform_int_distribution<uint32_t>(0, 1)(rng_);
-  auto frameTypeString = frameType(isRowsFrame);
+  }
+
+  if ((frameBoundType == core::WindowNode::BoundType::kFollowing &&
+       sortOrder.isAscending()) ||
+      (frameBoundType == core::WindowNode::BoundType::kPreceding &&
+       !sortOrder.isAscending())) {
+    if constexpr (std::is_same_v<T, double> || std::is_same_v<T, float>) {
+      return offsetColumnVal + offsetValue;
+    } else {
+      return checkedPlus<T>(offsetColumnVal, offsetValue);
+    }
+  }
+
+  VELOX_UNREACHABLE(
+      "Offset cannot be generated: sortOrder ascending {}, frameBoundType {}",
+      sortOrder.toString(),
+      core::WindowNode::boundTypeName(frameBoundType));
+}
+
+template <typename T>
+std::optional<std::string> WindowFuzzer::addKRangeFrameColumnToInput(
+    std::vector<RowVectorPtr>& input,
+    core::WindowNode::BoundType frameBoundType,
+    std::string& columnName,
+    SortingKeyAndOrder& orderByKey) {
+  T offsetValue;
+  if constexpr (!(std::is_same_v<T, bool> || std::is_same_v<T, StringView> ||
+                  std::is_same_v<T, Timestamp> ||
+                  std::is_same_v<T, UnknownValue>)) {
+    auto size = vectorFuzzer_.getOptions().vectorSize;
+    velox::test::VectorMaker vectorMaker{pool_.get()};
```

Comment:
Do we really need vectorMaker for the rowVector at the end ? It seems un-necessary.